Order specifications for hooks

rustyrussell opened the pull request:
This allows hooks to specify they must be before/after some other plugins.  Either by basename or full pathname.

As our plugin ecosystem grows and gets more sophisticated, such conflicts will inevitably occur where multiple plugins want to ensure they are called in certain order.  

We could implement a priority number system, but that's usually just window-dressing on what people really want (though, unlike this, would allow you to specify that you must be first, or last: we could add wildcards if we wanted to).

Getting this in now means that after a few releases, plugin authors will simply be able to assume its existence.

Fixes: #4005 (thanks @m-schmoock for pointing that out!)

m-schmoock commented:
This addresses https://github.com/ElementsProject/lightning/issues/4005

niftynei commented on lightningd/plugin.c:
```diff
@@ -1057,20 +1056,43 @@ static const char *plugin_subscriptions_add(struct plugin *plugin,
+					       json_tok_full_len(t),
+					       json_tok_full(buffer, t));
+			name = json_strdup(tmpctx, buffer, nametok);
+			beforetok = json_get_member(buffer, t, "before");
```
it took me a second to fully grasp the implication of 'before' and 'after'. Renaming them to "run_before" and "run_after" might be more immediately obvious?

rustyrussell replied:
It's more obvious in context, than when reading in a patch, e.g.:

```json
"hooks": [
    {"name": "htlc_accepted",
     "before": ["someplugin.py"],
     "after": ["foo.py"]}
]
```

niftynei reviewed (approved):
ACK 0c1d19b4c8ec4730a969551806224399faf0bae

m-schmoock commented:
concept ack

@rustyrussell  as you wrote in 4005
> 3\. Secondary sort by command line order, then config order.

I think it is also very helpful to give the user an option for plugin ordering (unless otherwise specified by the devs via before/after) by checking commandline and config order of plugins. The rationale behind is it the following: Before and after list are to be maintained by the plugin developers. Before that happens, a user typically has the problem first and needs to wait for the devs to to make the changes so his specific setup works. Since the devs will often not know all the plugin combinations out in the wild this can take a while or, in the worst case, never happens.

So I think we might also add the commandline/config order as 'last resort/default' if no before/after ordering has been specified for a plugin.

rustyrussell commented:
The current implementation has this effect, in fact. Technically it's not quite a stable sort, as dependencies can pull in things out of order, but we could fix that (I doubt anyone will notice, though).

For example, assume plugins in alphabetical order (they're in directory-sort order usually BTW). A has to be before D, and B has to be before C.

In this case, algorithm will go A B D C. Easy to fix though.

rustyrussell commented:
OK, fixed to make the default ordering equal to the specified one.

Also, removed a problem where we'd warn about missing dependencies since we evaluated them *as we loaded*.  This was more invasive, but I think simplifies the code a little anyway.

m-schmoock commented:
Thanks @rustyrussell 

Super wired, I stress tested your changes a little, but it seems there is an issue with the default commandline/config order (unless otherwise by before/after specified).

### How to reproduce
- check out ece24e5c9
- compile
- apply a test change, that now should work (see below)
- run the tests in parallel:  `py.test --count 50 -x -n 16 -v tests/test_plugin.py -k test_openchannel_hook_chaining`

### What happens
The flaky test can fail. When we look in the logs of l2 we see that the 'reject for a reason' hook is not called but 'reject by design', which should never happen if correct default order would be applied.

### The patch for the test that uses the new feature
```
diff --git a/tests/test_plugin.py b/tests/test_plugin.py
index d2628cd25..9f35a4dc5 100644
--- a/tests/test_plugin.py
+++ b/tests/test_plugin.py
@@ -606,24 +606,15 @@ def test_openchannel_hook_chaining(node_factory, bitcoind):
     l1, l2 = node_factory.line_graph(2, fundchannel=False, opts=opts)
     l1.fundwallet(10**6)
 
-    hook_msg = "openchannel_hook rejects and says '"
     # 100005sat fundchannel should fail fatal() for l2
     # because hook_accepter.py rejects on that amount 'for a reason'
     with pytest.raises(RpcError, match=r'They sent error channel'):
         l1.rpc.fundchannel(l2.info['id'], 100005)
 
-    # Note: hook chain order is currently undefined, because hooks are registerd
-    #       as a result of the getmanifest call, which may take some random time.
-    #       We need to workaround that fact, so test can be stable
-    correct_order = l2.daemon.is_in_log(hook_msg + "reject for a reason")
-    if correct_order:
-        assert l2.daemon.wait_for_log(hook_msg + "reject for a reason")
-        # the other plugin must not be called
-        assert not l2.daemon.is_in_log("reject on principle")
-    else:
-        assert l2.daemon.wait_for_log(hook_msg + "reject on principle")
-        # the other plugin must not be called
-        assert not l2.daemon.is_in_log("reject for a reason")
+    hook_msg = "openchannel_hook rejects and says '"
+    assert l2.daemon.wait_for_log(hook_msg + "reject for a reason")
+    # the third plugin must now not be called anymore
+    assert not l2.daemon.is_in_log("reject on principle")
 
     # 100000sat is good for hook_accepter, so it should fail 'on principle'
     # at third hook openchannel_reject.py
```

rustyrussell commented:
Argh, good catch: final patch actually fixes the problem, and for some reason I didn't push it!

rustyrussell commented:
And I turned your patch into a commit, credited you and appended. Thanks!

m-schmoock commented:
works like a charm

m-schmoock commented:
Regarding IRC chat:
> `00:22 < rusty> m-schmoock: how do you feel about getting rid of all those single hooks?  Or at least as many of them as possible...`

Can you elaborate on this?

cdecker commented on lightningd/plugin.c:
```diff
@@ -125,10 +147,16 @@ static void destroy_plugin(struct plugin *p)
-	/* Don't call this if we're still parsing options! */
-	if (p->plugin_state != UNCONFIGURED)
-		check_plugins_resolved(p->plugins);
+	/* If this was last one manifests were waiting for, handle deps */
+	if (p->plugin_state == AWAITING_GETMANIFEST_RESPONSE)
+		check_plugins_manifests(p->plugins);
+
+	/* If this was the last one init was waiting for, handle cmd replies */
+	if (p->plugin_state == AWAITING_INIT_RESPONSE)
+		check_plugins_initted(p->plugins);
```
This seems to be a weird place to perform these checks. We look at a random plugin being destroyed and if that was in a specific state we assume it must have been the last one in this state? We should kick this up to the set of plugins and perform the check there after removing the plugin to be destroyed. No need to replicate the last-one-turns-off-the-light test here :-)

rustyrussell replied:
No, we've already unlisted the plugin at this point.  We could, and probably should simply call check_plugins_xxx() unconditionally here.

I wonder if we should have a plugin_set_state() which calls these internally.  Less efficient, but clearer?

cdecker replied:
Similar to the `payment_continue` function that is used to continue in a state machine? I'm all for it ^^

cdecker commented on lightningd/plugin_hook.c:
```diff
@@ -434,6 +434,21 @@ static struct hook_node *find_hook(struct hook_node *graph, const char *name)
+		if ((*s)[pos]->hook->plugin->index > n->hook->plugin->index)
+			break;
+	}
+	tal_resize(s, s_num + 1);
+	memmove(*s + pos+1, *s + pos, sizeof(**s) * (s_num - pos));
+	(*s)[pos] = n;
```
I don't quite follow why this is working: we're striking dependencies off of the graph node until we finally find that it has no more, then add it to the list. However, before this change we added it to the end of the list, which was always safe (no dependency could jump over us). After this change we suddenly use insertion sort on the `index` at a potentially earlier position in the list, meaning we could end up being inserted before one of our dependencies. What am I missing here?

rustyrussell replied:
You're right!

We should instead gather all the things we are going to add from one edge, and order *those* then add as a group.  That means if B and C depend on A, we append A, then sort (B,C) and append those.

I will write a test which triggers this first...

cdecker commented on lightningd/plugin_hook.c:
```diff
@@ -392,3 +392,125 @@ void plugin_hook_add_deps(struct plugin_hook *hook,
+	for (size_t i = 0; i < tal_count(graph); i++) {
+		if (graph[i].num_incoming == 0)
+			tal_arr_expand(&s, &graph[i]);
+	}
```
This pre-populating is not necessary since it's just a special case for the below logic. The later topological sort will also make sure that this isn't required to maintain order for hooks without dependencies.

rustyrussell replied:
No, we need it because we only add the other candidates when we look at their count:

```
			if (--n->outgoing[i]->num_incoming == 0)
				add_to_s(&s, n->outgoing[i]);
```

Nodes which start with no incoming are not caught here.

cdecker replied:
Ah, gotcha, I had not seen the issue at first.

cdecker reviewed (commented):
Some minor questions for my own understanding, and a tiny python bug, otherwise a very solid PR :+1:

rustyrussell commented on contrib/pyln-client/pyln/client/plugin.py:
```diff
@@ -405,7 +407,9 @@ def decorator(f: Callable[..., JSONType]) -> Callable[..., JSONType]:
     def add_hook(self, name: str, func: Callable[..., JSONType],
-                 background: bool = False) -> None:
+                 background: bool = False,
+                 before: List[str] = [],
+                 after: List[str] = []) -> None:
```
Oh, nice catch, thanks!

rustyrussell commented:
I ended up reworking and thus simplifying the algo.  It's less efficient (O(N^2)) but nobody cares.

cdecker commented on contrib/pyln-client/pyln/client/plugin.py:
```diff
@@ -431,13 +431,15 @@ def add_hook(self, name: str, func: Callable[..., JSONType],
     def hook(self, method_name: str,
-             before: List[str] = [],
-             after: List[str] = []) -> JsonDecoratorType:
+             before: List[str] = None,
+             after: List[str] = None) -> JsonDecoratorType:
```
This now becomes an `Optional[List[str]]`

cdecker replied:
Sorry for forgetting this in my first round of comments :-)

cdecker commented on contrib/pyln-client/pyln/client/plugin.py:
```diff
@@ -431,13 +431,15 @@ def add_hook(self, name: str, func: Callable[..., JSONType],
         method.background = background
-        method.before = before
-        method.after = after
+        if before:
+            method.before = before
+        if after:
+            method.after = after
```
We should always set these, otherwise we'd have to check with `hasattr` whether the hook has that field.

```
method.before = [] if before else before
method.after = [] if after else after
```

rustyrussell replied:
Hmm, mypy didn't like that, so went for an open-coded variant.

cdecker commented on lightningd/plugin_hook.c:
```diff
@@ -401,25 +402,8 @@ void plugin_hook_add_deps(struct plugin_hook *hook,
- *     return L   (a topologically sorted order)
- */
 struct hook_node {
+	bool finished;
```
What are the semantics of `finished`? It's only used in sorting so I'd assume it means that it is in its final position.

rustyrussell replied:
Included in sorted results. I'll add a comment

cdecker commented on lightningd/plugin_hook.c:
```diff
@@ -434,31 +418,33 @@ static struct hook_node *find_hook(struct hook_node *graph, const char *name)
+		if (graph[i].num_incoming != 0)
+			continue;
+		if (!best
+		    || best->hook->plugin->index > graph[i].hook->plugin->index)
```
We're never moving the plugin in the `graph`, so under no circumstance should this ever be true (a later iteration should always have a higher `index`). Since that's the case I think we can remove the `index` altogether, and break/return after finding a first `best`

rustyrussell replied:
That's making an assumption that graph is in index order.  It (I'm fairly sure) starts that way, but can be perturbed by previous calls to plugin_hook_make_ordered.  If plugin B says it has to be before A and after C, we'll go C B A.  If B gets removed, we should return to A C.

Does anyone care?  Probably not, but this way is clear, at least.

rustyrussell replied:
On further thought, if we define it as "minimum perturbation to achieve required ordering" which is a totally reasonable goal, we can indeed do this simplification.

rustyrussell replied:
... Nope.  Hooks start in "who returned from getmanifest" order which is not the same as "order specified by config" order.

m-schmoock commented:
Hm, I restarted a travis flake which I couldn't reproduce locally also not by running it 100 times and parallel: `tests/test_plugin.py::test_plugin_command`

It was complaining about unclean node shutdown, but it succeeded upon first restart/retry.

m-schmoock commented:
@rustyrussell still (sorry for bugging):
> regarding IRC chat:
> 
> > `00:22 < rusty> m-schmoock: how do you feel about getting rid of all those single hooks?  Or at least as many of them as possible...`
> 
> Can you elaborate on this?
> 
> I was thinking that it would be sufficient to have the `before` and `after` lists on a per plugin level and not per hook. It would reduce (configuration) complexity. Having it on a per hook basis brings the option to have two different plugins with two (same) hooks each order their execution differently depending on what hook is called. This seems a bit overkill. Having plugin order rather than hook order would just prioritize one plugin over another, which will likely do the job.

rustyrussell commented:
> @rustyrussell still (sorry for bugging):
> 
> > regarding IRC chat:
> > > `00:22 < rusty> m-schmoock: how do you feel about getting rid of all those single hooks?  Or at least as many of them as possible...`
> > 
> > 
> > Can you elaborate on this?

We should make it so every hook can be registered by multiple plugins.

> > I was thinking that it would be sufficient to have the `before` and `after` lists on a per plugin level and not per hook. It would reduce (configuration) complexity. Having it on a per hook basis brings the option to have two different plugins with two (same) hooks each order their execution differently depending on what hook is called. This seems a bit overkill. Having plugin order rather than hook order would just prioritize one plugin over another, which will likely do the job.

I agree, at the moment.  But hooks are the only place where plugins interact, so I'd rather see explicitly what they care about.  If we ever get order conflicts, it's easier to see what they want when they have priorities on specific hooks.

rustyrussell commented:
Trivial rebase to fold in fixups and add a comment on `finished` var.

m-schmoock commented:
>>> 00:22 < rusty> m-schmoock: how do you feel about getting rid of all those single hooks? Or at least as many of them as possible...

>> Can you elaborate on this?

> We should make it so every hook can be registered by multiple plugins.

absolutely yes, I missed the point that you were referring to 'unchained' hooks.

m-schmoock commented on doc/PLUGINS.md:
```diff
@@ -678,14 +678,25 @@ declares that it'd like to be consulted on what to do next for certain
 react to the given event.
 
+When hooks are registered, they can optionally specify "before" and
+"after" arrays of plugin names, which control what order they will be
```
General question about 'plugin names': Is this same as for `cli plugin start/stop` commands, meaning either full path (useless for this feature) or basename? If so, we may note that the plugin name includes its extension, meaning full filename.

rustyrussell replied:
Yes, filename or full path name.

m-schmoock commented:
ACK 7d41997
